### PR TITLE
reset Metrics

### PR DESCRIPTION
**Pull request by donnut**
I needed a way to reset metrics, without removing the metric instances, see also #145. 
This solution is not completely tested, but I would like to know if this is a right way to go.

### CODE REVIEW

File: test/registerTest.js [hunk 1]
```diff
@@ -204,6 +204,49 @@ describe('register', () => {
 		expect(output).toEqual(metric);
 	});
 
+	it('should allow resetting a counter', () => {
```
Comment (SimenB) [inline test/registerTest.js]: the tests should be in their respective as well

File: test/registerTest.js [hunk 2]
```diff
@@ -204,6 +204,49 @@ describe('register', () => {
...
+			labelNames: ['serial', 'active']
+		});
+		register.registerMetric(metric);
+
+		metric.inc({ serial: '12345', active: 'yes' }, 12);
+		expect(metric.get().values[0].value).toEqual(12);
+		expect(metric.get().values[0].labels.serial).toEqual('12345');
+		expect(metric.get().values[0].labels.active).toEqual('yes');
+
+		register.resetMetrics('serial');
```
Comment (SimenB) [inline test/registerTest.js]: `'serial'`? Doesn't seem like the method takes an argument?

File: lib/registry.js [hunk 3]
```diff
@@ -101,6 +101,12 @@ class Registry {
 		this._defaultLabels = labels;
 	}
 
+	resetMetrics() {
+		for (const metric in this._metrics) {
```
Comment (SimenB) [inline lib/registry.js]: `this._metrics.forEach(...`

File: test/registerTest.js (hunk 1 above)
Comment (SimenB) [reply test/registerTest.js]: `registerTest` can have a single tests with all metric types for the coverage 🙂

File: lib/counter.js [hunk 4]
```diff
@@ -116,6 +124,14 @@ class Counter {
 	}
 }
 
+const reset = function() {
+	this.hashMap = {};
```
Comment (siimon) [inline lib/counter.js]: Might be a good idea to call reset in the constructor instead of having duplicate code

File: lib/gauge.js [hunk 5]
```diff
@@ -226,6 +234,14 @@ function set(labels) {
 	};
 }
 
+function reset() {
+	this.hashMap = {};
+
```
Comment (siimon) [inline lib/gauge.js]: Same as in the counter, call this function from the constructor

File: test/counterTest.js [hunk 6]
```diff
@@ -240,4 +240,42 @@ describe('counter', () => {
 			expect(instance.get().values[0].timestamp).toEqual(undefined);
 		});
 	});
+	describe('counter reset', () => {
+		xit('should reset labelless counter', () => {
```
Comment (siimon) [inline test/counterTest.js]: Why is this test disabled?

File: test/gaugeTest.js [hunk 7]
```diff
@@ -375,6 +375,44 @@ describe('gauge', () => {
 			});
 		});
 	});
+	describe('gauge reset', () => {
+		xit('should reset labelless gauge', () => {
```
Comment (siimon) [inline test/gaugeTest.js]: Disabled test again, enable or remove :)

File: README.md [hunk 8]
```diff
@@ -104,6 +104,12 @@ counter.inc(); // Inc with 1
 counter.inc(10); // Inc with 10
 ```
 
+A counter can be reset manually.
```
Comment (siimon) [inline README.md]: Maybe we could add what reset means, ie removing all label-value combinations and initialize to 0

Comment (donnut) [reply README.md]: Good point. I'll extend the text.

File: lib/counter.js (hunk 4 above)
Comment (donnut) [reply lib/counter.js]: Ok

File: test/counterTest.js (hunk 6 above)
Comment (donnut) [reply test/counterTest.js]: By accident.

File: test/gaugeTest.js (hunk 7 above)
Comment (donnut) [reply test/gaugeTest.js]: Same thing

**Comment by donnut** [comment]
@SimenB I'm not that familair with `Histogram` and `Summary` yet, but I was under the impression that the current `reset` for both are already implemented. I added a test in registryTest that call their reset methods.

**Comment by SimenB** [comment]
Oh, would you look at that, haha! It's not documented though, would you mind adding it to the readme? 🙂

**Comment by siimon** [comment]
Thanks! 👍

**Comment by SimenB** [comment]
Released as 10.2.0

**Pull request merged**